### PR TITLE
Refine decision event logic

### PR DESCRIPTION
**Pull request by lihuoran**
# Description

- Add `DecisionEventPayload` and `ActionPayload` as the base class for payloads of decision events & action events.
- Add related type check during running.
- Rename related variables under `simulator/`.
- Refine the action handling logic. Modify `Env` and related BEs.

## Linked issue(s)/Pull request(s)

- [issue_number](issue_link)

## Type of Change

- [ ] Non-breaking bug fix
- [ ] Breaking bug fix
- [ ] New feature
- [ ] Test
- [ ] Doc update
- [ ] Docker update

## Related Component

- [ ] Simulation toolkit
- [ ] RL toolkit
- [ ] Distributed toolkit

## Has Been Tested

- OS:
  - [ ] Windows
  - [ ] Mac OS
  - [ ] Linux
- Python version:
  - [ ] 3.7
  - [ ] 3.8
  - [ ] 3.9
- Key information snapshot(s):

## Needs Follow Up Actions

- [ ] New release package
- [ ] New docker image

## Checklist

- [ ] Add/update the related comments
- [ ] Add/update the related tests
- [ ] Add/update the related documentations
- [ ] Update the dependent downstream modules usage

### CODE REVIEW

File: maro/simulator/scenarios/cim/business_engine.py [hunk 1]
```diff
@@ -705,47 +705,47 @@ def _on_empty_return(self, event: AtomEvent):
         port.on_consignee -= payload.quantity
         port.empty += payload.quantity
 
-    def _on_action_received(self, event: CascadeEvent):
+    def _on_action_received(self, event: ActualEvent):
```
Comment (Jinyu-W) [inline maro/simulator/scenarios/cim/business_engine.py:708]: the event for action is a CascadeEvent, never be an AtomEvent

File: maro/simulator/scenarios/citi_bike/business_engine.py [hunk 2]
```diff
@@ -519,20 +519,21 @@ def _on_bike_deliver(self, evt: AtomEvent):
 
         station.bikes = station_bikes + max_accept_number
 
-    def _on_action_received(self, evt: AtomEvent):
+    def _on_action_received(self, evt: ActualEvent):
```
Comment (Jinyu-W) [inline maro/simulator/scenarios/citi_bike/business_engine.py:522]: CacadeEvent

File: maro/simulator/scenarios/citi_bike/business_engine.py [hunk 3]
```diff
@@ -519,20 +519,21 @@ def _on_bike_deliver(self, evt: AtomEvent):
...
+            assert isinstance(action, Action)
 
-        for action in evt.payload:
             from_station_idx: int = action.from_station_idx
             to_station_idx: int = action.to_station_idx
 
             # Ignore invalid cell idx.
             if from_station_idx < 0 or to_station_idx < 0:
-                continue
+                return
```
Comment (Jinyu-W) [inline maro/simulator/scenarios/citi_bike/business_engine.py:536]: why return instead of continue?

File: tests/cim/test_cim_scenario.py [hunk 4]
```diff
@@ -393,7 +393,7 @@ def test_early_discharge(self) -> None:
             self._init_env(backend_name)
 
             metric, decision_event, is_done = self._env.step(None)
```
Comment (Jinyu-W) [inline tests/cim/test_cim_scenario.py:395]: rename `decision_event` to `decision_payload`?

File: maro/simulator/core.py [hunk 5]
```diff
@@ -292,50 +344,23 @@ def _simulate(self) -> Generator[Tuple[dict, List[object], bool], object, None]:
...
-                decision_events = [event.payload for event in pending_events]
-
-                decision_events = (
-                    decision_events[0] if self._decision_mode == DecisionMode.Sequential else decision_events
-                )
-
-                # Yield current state first, and waiting for action.
-                actions = yield self._business_engine.get_metrics(), decision_events, False
-                # archive decision events.
-                self._decision_events.append(decision_events)
```
[the remark this answers is not in the document]
Comment (lihuoran) [reply maro/simulator/core.py:304]: Fixed

File: maro/simulator/scenarios/cim/business_engine.py (hunk 1 above)
Comment (lihuoran) [reply maro/simulator/scenarios/cim/business_engine.py:708]: Fixed. Also for CitiBike.

File: maro/common.py [hunk 6]
```diff
@@ -0,0 +1,21 @@
...
+
+
+class BaseDecisionEvent:
+    """Base class for all decision events.
+
+    We made this design for the convenience of users. As a summary, there are two types of events in MARO:
+    - CascadeEvent & AtomEvent: used to drive the MARO Env / business engine.
+    - DecisionEvent: exposed to users as a means of communication.
+
+    The latter one serves the `payload` of the former ones inside of MARO Env.
```
Comment (Jinyu-W) [inline maro/common.py:12]: `serves as`?

File: maro/simulator/core.py [hunk 7]
```diff
@@ -89,21 +89,48 @@ def __init__(
...
 
-    def step(self, action) -> Tuple[Optional[dict], Optional[List[object]], Optional[bool]]:
+    def step(
+        self,
+        action: Union[BaseAction, List[BaseAction], None] = None,
+    ) -> Tuple[Optional[dict], Union[BaseDecisionEvent, List[BaseDecisionEvent], None], bool]:
         """Push the environment to next step with action.
 
+        Under Sequential mode:
+            - If `action` is None, a empty list will be assigned to the decision event.
```
Comment (Jinyu-W) [inline maro/simulator/core.py:99]: `an empty list`

File: maro/simulator/core.py [hunk 8]
```diff
@@ -89,21 +89,48 @@ def __init__(
...
+            Will assign `a1` & `a2` to the first decision event, `a3` to the second decision event, and `a4` & `a5`
+            to the third decision event.
+
+            Particularly, if you only want to assign multiple actions to the first decision event, please
+            pass `[[a1, a2, ..., an]]` (a list of one list) instead of `[a1, a2, ..., an]` (an 1D list of n elements),
+            since the latter one will assign the n actions to the first n decision events.
+
         Args:
-            action (Action): Action(s) from agent.
+            action (Union[BaseAction, List[ActionPayload], None]): Action(s) from agent.
```
Comment (Jinyu-W) [inline maro/simulator/core.py:123]: what's the difference between `BaseAction` and `ActionPayload`?

Comment (lihuoran) [reply maro/simulator/core.py:123]: Typo here.

**Review by Jinyu-W — APPROVED** [review]

**Pull request merged**